### PR TITLE
ADBDEV-6682: Implement TLS options for PXF external tables

### PR DESCRIPTION
**Pull request by andr-sokolov**
TLS options implemented for PXF external tables

The requirement for using TLS is implemented for PXF external tables.
Communication takes place between the segment node and the PXF server and is
pointless without the PXF server. Testing should be done in conjunction with PXF
server changes and is out of scope for this commit.

PXF external tables are controlled by environment variables. The following
options are affected by this commit:

* PXF_PROTOCOL: Type string, default http. Defines the protocol for
  communication with the PXF server. Possible values are http and https.
* PXF_SSL_CACERT: Type string, default
  /home/gpadmin/arenadata_configs/cacert.pem. Specifies the absolute path to the
  CA certificate that signed the server certificate. Refers to CURLOPT_CAINFO.
* PXF_SSL_CERT: Type string, default client.pem. Specifies the path to the
  client certificate. Refers to CURLOPT_SSLCERT.
* PXF_SSL_CERT_TYPE: Type string, default pem. Defines the type of the client
  certificate. Refers to CURLOPT_SSLCERTTYPE.
* PXF_SSL_KEY: Type string, default empty. Defines the path to the client's SSL
  private key. Refers to CURLOPT_SSLKEY.
* PXF_SSL_KEYPASSWD: Type string, default empty. Defines the password for the
  client SSL private key. Refers to CURLOPT_KEYPASSWD.
* PXF_SSL_VERIFY_PEER: Type long, default 1. Defines whether the client should
  verify the server certificate. Refers to CURLOPT_SSL_VERIFYPEER.

### CODE REVIEW

File: external-table/src/libchurl.c [hunk 1]
```diff
@@ -415,11 +416,43 @@ churl_init(const char *url, CHURL_HEADERS headers)
...
 
 	return (CHURL_HANDLE) context;
 }
 
+static void
+set_curl_ssl_options(churl_context *context)
+{
+	const char *proto = get_pxf_protocol();
+
+	if (proto && strcmp(proto, "https") == 0 )
```
Comment (andr-sokolov) [inline external-table/src/libchurl.c]:
```suggestion
	if (proto && strcmp(proto, "https") == 0) 
```

File: external-table/src/pxfutils.c [hunk 2]
```diff
@@ -76,6 +76,18 @@ get_authority(void)
 	return psprintf("%s:%d", get_pxf_host(), get_pxf_port());
 }
 
+const char *
+get_pxf_protocol(void)
+{
+	const char *proto = getenv(ENV_PXF_PROTOCOL);
+
+	if (proto == NULL) {
```
Comment (andr-sokolov) [inline external-table/src/pxfutils.c]: An opening bracket after `if` should be in the next line - this is code style of PostgreSQL. As an alternative the brackets can be removed here

File: external-table/src/pxfutils.c [hunk 3]
```diff
@@ -116,6 +128,85 @@ get_pxf_port(void)
...
+{
+	const char *verify_peer_var = getenv(ENV_PXF_SSL_VERIFY_PEER);
+	long verifypeer = PXF_DEFAULT_SSL_VERIFY_PEER; 
+
+	if (verify_peer_var != NULL) 
+	{
+		verifypeer = atol(verify_peer_var);
+	}
+
+	return verifypeer;
```
Comment (andr-sokolov) [inline external-table/src/pxfutils.c]: I suggest to remove the verifypeer variable
```suggestion

	if (verify_peer_var != NULL) 
	{
		return atol(verify_peer_var);
	}

	return PXF_DEFAULT_SSL_VERIFY_PEER;
```

File: external-table/src/pxfutils.c [hunk 4]
```diff
@@ -116,6 +128,85 @@ get_pxf_port(void)
 	return port;
 }
 
+const char *
+get_pxf_ssl_keypasswd(void)
```
Comment (RekGRpth) [inline external-table/src/pxfutils.c:145]: how about only two functions
```c
const char *getenv_char(const char *name, const char *default);
long getenv_long(const char *name, long default);
```

Comment (dkovalev1) [reply external-table/src/pxfutils.c:145]: Nice idea, implemented.

File: external-table/src/pxfutils.c (hunk 3 above)
Comment (dkovalev1) [reply external-table/src/pxfutils.c]: This code rewritten now.

File: external-table/src/pxfutils.c (hunk 2 above)
Comment (dkovalev1) [reply external-table/src/pxfutils.c]: okay

File: external-table/src/libchurl.c (hunk 1 above)
Comment (dkovalev1) [reply external-table/src/libchurl.c]: okay

File: external-table/src/pxfutils.c (hunk 4 above)
Comment (RekGRpth) [reply external-table/src/pxfutils.c:145]: Actually, I meant to get rid of the new `get_pxf_ssl_*` functions completely, and instead just use the two `getenv_char` and `getenv_long` functions.

File: external-table/src/pxfutils.c [hunk 5]
```diff
@@ -69,13 +72,39 @@ concat(int num_args,...)
...
 }
 
+static const char*
+getenv_char(const char *name, const char *default_value)
+{
+	const char *value = getenv(name);
+	if (value == NULL)
+		return default_value;
+	
+	return value;
```
Comment (RekGRpth) [inline external-table/src/pxfutils.c]: why not
```suggestion
	return value != NULL ? value : default_value;
```

File: external-table/src/pxfutils.c (hunk 4 above)
Comment (dkovalev1) [reply external-table/src/pxfutils.c:145]: get_pxf_ssl_* functions keep PXF-related setup isolated in the pxfutils.c, continuing tradition started with get_pxf_port(), get_authority(), etc. It keeps code less coupled.
If fact I'd rename get_pxf_ssl_* functions to get_ssl_* to make churl logically less dependent from PXF, what do you think?

File: external-table/src/pxfutils.c (hunk 5 above)
Comment (dkovalev1) [reply external-table/src/pxfutils.c]: > why not

Because ternary operator is less clear for the folks, as you taught me previously.

Comment (RekGRpth) [reply external-table/src/pxfutils.c]: see `get_pxf_host`

File: external-table/src/pxfutils.h [hunk 6]
```diff
@@ -25,17 +28,39 @@ const char *get_pxf_host(void);
...
+#define ENV_PXF_SSL_CERT_TYPE       "PXF_SSL_CERT_TYPE"
+#define ENV_PXF_SSL_KEY             "PXF_SSL_KEY"
+#define ENV_PXF_SSL_KEYPASSWD       "PXF_SSL_KEYPASSWD"
+#define ENV_PXF_SSL_VERIFY_PEER     "PXF_SSL_VERIFY_PEER"
+#define PXF_DEFAULT_HOST            "localhost"
+#define PXF_DEFAULT_PORT            5888
+#define PXF_DEFAULT_PROTOCOL        "http"
+#define PXF_DEFAULT_SSL_CACERT      "/home/gpadmin/arenadata_configs/cacert.pem"
+#define PXF_DEFAULT_SSL_CERT        "client.p12"
+#define PXF_DEFAULT_SSL_CERT_TYPE   "P12"
```
Comment (KnightMurloc) [inline external-table/src/pxfutils.h]: According to the specification, these parameters should be "client.pem" and "pem".

File: external-table/src/pxfutils.c [hunk 7]
```diff
@@ -69,13 +72,39 @@ concat(int num_args,...)
...
+}
+
+static long 
+getenv_long(const char *name, long default_value)
+{
+	const char *value = getenv(name);
+	if (value == NULL)
+		return default_value;
+
+	return atol(value);
```
Comment (KnightMurloc) [inline external-table/src/pxfutils.c]: do we need to check the correctness of the conversion? Maybe it's better to use `strtol` and check for errors?

File: external-table/src/libchurl.c [hunk 8]
```diff
@@ -415,11 +416,42 @@ churl_init(const char *url, CHURL_HEADERS headers)
...
+		{
+			set_curl_option(context, CURLOPT_SSLKEYPASSWD, keypasswd);
+		}
+
+		if (cacert != NULL && cacert[0] != '\0') 
+		{
+			set_curl_option(context, CURLOPT_CAINFO, cacert);
+		}
+
+		set_curl_option(context, CURLOPT_SSL_VERIFYPEER, (const void *)get_pxf_ssl_verifypeer());
```
Comment (KnightMurloc) [inline external-table/src/libchurl.c]:
```suggestion
		set_curl_option(context, CURLOPT_SSL_VERIFYPEER, (const void *) get_pxf_ssl_verifypeer());
```

File: external-table/src/libchurl.c [hunk 9]
```diff
@@ -415,11 +416,42 @@ churl_init(const char *url, CHURL_HEADERS headers)
...
 
 	return (CHURL_HANDLE) context;
 }
 
+static void
+set_curl_ssl_options(churl_context *context)
+{
+	const char *proto = get_pxf_protocol();
+
+	if (proto && strcmp(proto, "https") == 0)
```
Comment (KnightMurloc) [inline external-table/src/libchurl.c:432]: It looks like `proto`, `keypasswd` and `cacert` cannot be NULL. Do we need a check them for NULL?

File: external-table/src/pxfutils.c (hunk 7 above)
Comment (dkovalev1) [reply external-table/src/pxfutils.c]: atol provides meaningful value of 0 when invalid input provided, and I believe it's quite clear for the user.

File: external-table/src/libchurl.c (hunk 9 above)
Comment (dkovalev1) [reply external-table/src/libchurl.c:432]: Right, in the current implementation it's not NULL, but this check is a good practice for the case if someone changes the behavior get_pxf_protocol(). And it's cheap and readable.

File: external-table/src/libchurl.c (hunk 8 above)
Comment (dkovalev1) [reply external-table/src/libchurl.c]: ok

File: external-table/src/pxfutils.h (hunk 6 above)
Comment (dkovalev1) [reply external-table/src/pxfutils.h]: Right, spec has changed afterwards. Fixed.

File: external-table/src/libchurl.c [hunk 10]
```diff
@@ -415,11 +416,42 @@ churl_init(const char *url, CHURL_HEADERS headers)
...
+	{
+		const char *keypasswd = get_pxf_ssl_keypasswd();
+		const char *cacert = get_pxf_ssl_cacert(); // PXF_SSL_CACERT
+
+		set_curl_option(context, CURLOPT_SSLCERT, get_pxf_ssl_cert());
+		set_curl_option(context, CURLOPT_SSLKEY, get_pxf_ssl_key());
+
+		set_curl_option(context, CURLOPT_SSLCERTTYPE, get_pxf_ssl_certtype());
+		if (keypasswd != NULL)
```
Comment (KnightMurloc) [inline external-table/src/libchurl.c]:
```suggestion
		if (keypasswd != NULL)
```

File: external-table/src/libchurl.c [hunk 11]
```diff
@@ -415,11 +416,42 @@ churl_init(const char *url, CHURL_HEADERS headers)
...
+		set_curl_option(context, CURLOPT_SSLCERT, get_pxf_ssl_cert());
+		set_curl_option(context, CURLOPT_SSLKEY, get_pxf_ssl_key());
+
+		set_curl_option(context, CURLOPT_SSLCERTTYPE, get_pxf_ssl_certtype());
+		if (keypasswd != NULL) 
+		{
+			set_curl_option(context, CURLOPT_SSLKEYPASSWD, keypasswd);
+		}
+
+		if (cacert != NULL && cacert[0] != '\0')
```
Comment (KnightMurloc) [inline external-table/src/libchurl.c]:
```suggestion
		if (cacert != NULL && cacert[0] != '\0')
```

File: external-table/src/pxfbridge.c [hunk 12]
```diff
@@ -265,8 +265,9 @@ build_uri_for_read(gphadoop_context *context)
 static void
 build_uri_for_write(gphadoop_context *context)
 {
-	appendStringInfo(&context->uri, "http://%s/%s/write",
-					 get_authority(), PXF_SERVICE_PREFIX);
+
+	appendStringInfo(&context->uri, "%s://%s/%s/write",
+					 get_pxf_protocol(), get_authority(), PXF_SERVICE_PREFIX);
```
Comment (KnightMurloc) [inline external-table/src/pxfbridge.c]: An empty line appeared.
```suggestion
	appendStringInfo(&context->uri, "%s://%s/%s/write",
					 get_pxf_protocol(), get_authority(), PXF_SERVICE_PREFIX);
```

**Review by KnightMurloc — APPROVED** [review]

File: external-table/src/pxfutils.h [hunk 13]
```diff
@@ -25,17 +28,39 @@ const char *get_pxf_host(void);
...
+#define ENV_PXF_SSL_VERIFY_PEER     "PXF_SSL_VERIFY_PEER"
+#define PXF_DEFAULT_HOST            "localhost"
+#define PXF_DEFAULT_PORT            5888
+#define PXF_DEFAULT_PROTOCOL        "http"
+#define PXF_DEFAULT_SSL_CACERT      "/home/gpadmin/arenadata_configs/cacert.pem"
+#define PXF_DEFAULT_SSL_CERT        "client.pem"
+#define PXF_DEFAULT_SSL_CERT_TYPE   "pem"
+#define PXF_DEFAULT_SSL_KEY         ""
+#define PXF_DEFAULT_SSL_KEYPASSWD   ""
+#define PXF_DEFAULT_SSL_VERIFY_PEER 1
```
Comment (andr-sokolov) [inline external-table/src/pxfutils.h:63]: Can we don't add new macros, but write their values as arguments of getenv_char and getenv_long? These macros are use only once

Comment (dkovalev1) [reply external-table/src/pxfutils.h:63]: These are according to the spec, and also it's a convenient place where all values gathered together.

File: external-table/src/libchurl.c [hunk 14]
```diff
@@ -415,11 +416,42 @@ churl_init(const char *url, CHURL_HEADERS headers)
...
 
+static void
+set_curl_ssl_options(churl_context *context)
+{
+	const char *proto = get_pxf_protocol();
+
+	if (proto && strcmp(proto, "https") == 0)
+	{
+		const char *keypasswd = get_pxf_ssl_keypasswd();
+		const char *cacert = get_pxf_ssl_cacert(); // PXF_SSL_CACERT
```
Comment (andr-sokolov) [inline external-table/src/libchurl.c]: Can we remove this comment? I think that the function name explains more than the comment

Comment (dkovalev1) [reply external-table/src/libchurl.c]: done

File: external-table/src/pxfutils.h [hunk 15]
```diff
@@ -25,17 +28,39 @@ const char *get_pxf_host(void);
...
+
```
Comment (andr-sokolov) [inline external-table/src/pxfutils.h:64]: Redundant empty line is added. Other header files in this directory contain only one empty line before "#endif"

File: external-table/src/libchurl.c [hunk 16]
```diff
@@ -415,11 +416,42 @@ churl_init(const char *url, CHURL_HEADERS headers)
...
+		set_curl_option(context, CURLOPT_SSLCERT, get_pxf_ssl_cert());
+		set_curl_option(context, CURLOPT_SSLKEY, get_pxf_ssl_key());
+
+		set_curl_option(context, CURLOPT_SSLCERTTYPE, get_pxf_ssl_certtype());
+		if (keypasswd != NULL)
+		{
+			set_curl_option(context, CURLOPT_SSLKEYPASSWD, keypasswd);
+		}
+
+		if (cacert != NULL && cacert[0] != '\0')
```
Comment (andr-sokolov) [inline external-table/src/libchurl.c:442]: Values for some option are checked before setting, but some values are not checked. What is the difference between options?

Comment (andr-sokolov) [inline external-table/src/libchurl.c:442]: Why do we need these checks for CURLOPT_CAINFO? It is written in the documentation: "Set it to NULL to disable its use again" (https://curl.se/libcurl/c/CURLOPT_CAINFO.html)
I tried do set the option as NULL and "", and didn't get segfault
```c
#include <curl/curl.h>

int main(void)
{
  CURL *curl = curl_easy_init();
  if(curl) {
    curl_easy_setopt(curl, CURLOPT_URL, "https://ya.ru/");
    curl_easy_setopt(curl, CURLOPT_CAINFO, NULL);
    curl_easy_perform(curl);
    curl_easy_cleanup(curl);
  }
}
```

File: external-table/src/libchurl.c [hunk 17]
```diff
@@ -415,11 +416,42 @@ churl_init(const char *url, CHURL_HEADERS headers)
...
+		const char *keypasswd = get_pxf_ssl_keypasswd();
+		const char *cacert = get_pxf_ssl_cacert();
+
+		set_curl_option(context, CURLOPT_SSLCERT, get_pxf_ssl_cert());
+		set_curl_option(context, CURLOPT_SSLKEY, get_pxf_ssl_key());
+
+		set_curl_option(context, CURLOPT_SSLCERTTYPE, get_pxf_ssl_certtype());
+		if (keypasswd != NULL)
+		{
+			set_curl_option(context, CURLOPT_SSLKEYPASSWD, keypasswd);
```
Comment (andr-sokolov) [inline external-table/src/libchurl.c]: The spec says: "PXF_SSL_KEYPASSWD ... It refers to [CURLOPT_KEYPASSWD](https://curl.se/libcurl/c/CURLOPT_KEYPASSWD.html).", but CURLOPT_SSLKEYPASSWD is set

File: external-table/src/pxfutils.h (hunk 15 above)
Comment (dkovalev1) [reply external-table/src/pxfutils.h:64]: removed

File: external-table/src/libchurl.c (hunk 16 above)
Comment (dkovalev1) [reply external-table/src/libchurl.c:442]: Removed check for keypasswd as default values will be used, which is not NULL. Check for cacert is a requirement from the SRS

File: external-table/src/libchurl.c (hunk 17 above)
Comment (dkovalev1) [reply external-table/src/libchurl.c]: The are the same but CURLOPT_KEYPASSWD is more recent. Changed to it.

**Comment by andr-sokolov** [comment]
Approved. I created the PR, so I cannot approve it in the standard way

**Pull request merged**